biodatasets/osiris/osiris.py

```diff
@@ -0,0 +1,302 @@
+                },
+            )
+        ]
+
+    def _get_offsets(self, parent: ET.Element, child: ET.Element) -> List[int32]:
+        '''
+        Retrieves character offsets for child from parent.
+        '''
+        parent_text = " ".join([" ".join([t for t in c.itertext()])
+                               for c in list(parent) if c.tag != "Pmid"])
+        child_text = " ".join([t for t in child.itertext()])
+        start = parent_text.index(child_text)
+        end = start + len(child_text)
+        return [start, end]
+
+    def _get_dict(self, elem: ET.Element) -> dict:
+        '''
+        Retrieves dict from XML element.
+        '''
+        elem_d = dict()
+        for child in elem:
+            elem_d[child.tag] = {}
+            elem_d[child.tag]["text"] = " ".join([t for t in child.itertext()])
+
+            if child.tag != "Pmid":
+                elem_d[child.tag]["offsets"] = self._get_offsets(elem, child)
+
+            for c in child:
+                elem_d[c.tag] = []
+
+            for c in child:
+                c_dict = c.attrib
+                c_dict["offsets"] = self._get_offsets(child, c)
+                elem_d[c.tag].append(c.attrib)
+
+        return elem_d
+
+    def _handle_missing_variants(self, row: dict) -> dict:
+        '''
+        If variant is not present in the row this function adds one variant 
+        with no data (to make looping though items possible) and returns the new row.
+        Otherwise returns unchanged row.
+        '''
+
+        if row.get("variant", 0) == 0:
+            row["variant"] = [
+                {
+                    "v_id": "",
+                    "v_lex": "",
+                    "offsets": [0, 0]
+                }
+            ]
+        return row
+
+    def _get_entities(self, row: dict) -> List[dict]:
+        '''
+        Retrieves two lists of dicts for genes and variants. 
+        After that, chains both together.
+        '''
+        genes = [
+            {
+                "id": str(uuid.uuid4()),
+                "offsets": [gene["offsets"]],
+                "text": [" ".join([gene["g_id"], gene["g_lex"]])],
+                "type": "gene",
+                "normalized": []
+            } for gene in row["gene"]
+        ]
+        variants = [
+            {
+                "id": str(uuid.uuid4()),
+                "offsets": [variant["offsets"]],
+                "text": [" ".join([variant["v_id"], variant["v_lex"]])],
+                "type": "variant",
+                "normalized": []
+            } for variant in row["variant"]
+        ]
+        return list(itertools.chain(genes, variants))
+
+    def _generate_examples(self, filepath, split):
+
+        root = ET.parse(filepath).getroot()
+        uid = 0
+        if self.config.schema == "source":
+            for elem in list(root):
+                row = self._get_dict(elem)
+
+                # handling missing variants data
+                row = self._handle_missing_variants(row)
```

Comment:
If I understand correctly, this adds "mock" entities to facilitate parsing. Could you please make sure that you remove them after parsing is completed? If I am misunderstanding this please feel free to ignore this comment

Reply:
Thank you for your suggestions and help! Yes, I've made it specifically for parsing. I add a line to my code that makes sure that these are getting removed after parsing procedures.